shared/js/content-scope/canvas.js

```diff
@@ -17,38 +17,86 @@ export function computeOffScreenCanvas (canvas, domainKey, sessionKey, getImageD
 }
 
 export function modifyPixelData (imageData, domainKey, sessionKey) {
-    const arr = []
-    // We calculate a checksum as passing imageData as a key is too slow.
-    // We might want to do something more pseudo random that is less observable through timing attacks and collisions (but this will come at a performance cost)
-    let checkSum = 0
-    // Create an array of only pixels that have data in them
-    const d = imageData.data
-    for (let i = 0; i < d.length; i += 4) {
-        // Blank Blue and Green color
-        if (d[i + 1] === 0 && d[i + 2] === 0) {
-            // Ignore non blank pixels there is high chance compression ignores them
-            if (d[i] === 0 && d[i + 3] === 0) {
-                continue
+    const length = imageData.data.length / 4
+    const hitMinimum = 50
+    const hitMax = 500
+    const windowSize = 2000
+    const windows = Math.ceil(length / windowSize)
+    const remainder = length % windowSize
+    for (let windowNumber = 0; windowNumber < windows; windowNumber++) {
+        const windowStartIndex = windowNumber * windowSize
+        let windowLength = windowSize
+        if (windowNumber === windows - 1) {
+            windowLength = remainder
+        }
+        const checksum = getChecksum(imageData.data, windowStartIndex, windowStartIndex + windowLength)
+        const windowHash = getDataKeySync(sessionKey, domainKey, checksum + windowNumber)
+
+        let hits = 0
+        iterateDataKey(windowHash, (item, byte) => {
+            const channel = byte % 3
+            const lookupId = item % windowLength
+            const pixelCanvasIndex = windowStartIndex + lookupId + channel
+            if (shouldIgnorePixel(imageData.data, pixelCanvasIndex)) {
+                return
             }
-            // Ignore phaser background
-            if (d[i] === 255 && d[i + 3] === 255) {
-                continue
+
+            ++hits
+            imageData.data[pixelCanvasIndex] = imageData.data[pixelCanvasIndex] ^ (byte & 0x1)
+        })
+
+        if (hits < hitMinimum) {
```

Comment:
I wonder whether this is necessary? I suspect that most fingerprinting scripts will simply hash the resulting canvas, so as long as we are modifying some pixels in every canvas the protection should be effective.

Reply:
I can see how this is necessary for certain types of canvases. See my comment here: https://app.asana.com/0/0/1200763289112611/1200838871508698/f for more thoughts.